### DISCUSSION TITLE
沟槽的默认弹幕云屏蔽

### DISCUSSION THREAD

**Discussion by KuiPer-LTW** [the1812/Bilibili-Evolved#4828] [功能建议 / Ideas]
我自己有屏蔽词不需要它帮我屏蔽，搞的现在很多视频明明有几千条弹幕但却显示不了几条

希望能有个取消云屏蔽的功能

![云屏蔽](https://github.com/user-attachments/assets/3dc52f26-2044-4d92-ad10-6e51cebca1d6)

**Comment by WakelessSloth56** [comment]
这个也是灰度测试的内容，目前改一下 `webAbTest` 还能关，但以后应该会全面应用的……
```
unsafeWindow.webAbTest.danmuku_block_version = "";
```

**Comment by the1812** [comment]
怎么会有这种迷惑功能。。。

**Comment by festoney8** [comment]
是灰测，会直接把禁用弹幕等级屏蔽的勾选框给你干掉

![image](https://github.com/user-attachments/assets/d5c1bc7d-1116-47b3-a133-ff46a981bc89)

还是用脚本解决，但不知道能撑多久

```JavaScript
// ==UserScript==
// @name         禁用弹幕屏蔽灰测
// @namespace    http://tampermonkey.net/
// @version      2024-07-07
// @description  禁用弹幕屏蔽灰测
// @author       You
// @match        https://www.bilibili.com/video/*
// @icon         https://www.google.com/s2/favicons?sz=64&domain=bilibili.com
// @grant        none
// @run-at       document-start
// ==/UserScript==
(function () {
    "use strict";
    let origValue = window.webAbTest;
    if (origValue) {
        origValue.danmuku_block_version = "OLD";
    }
    Object.defineProperty(window, "webAbTest", {
        get() {
            return origValue;
        },
        set(value) {
            if (value) {
                value.danmuku_block_version = "OLD";
            }
            origValue = value;
        }
    });
})();
```

**Reply by KuiPer-LTW** [reply]
谢谢，有用

**Reply by KuiPer-LTW** [reply]
似乎GG了
![弹幕设置](https://github.com/user-attachments/assets/322dbc9a-a485-4713-9a94-265bf4359a38)

**Reply by festoney8** [reply]
> 似乎GG了 ![弹幕设置](https://private-user-images.githubusercontent.com/103661658/358967473-322dbc9a-a485-4713-9a94-265bf4359a38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwMzg4ODksIm5iZiI6MTcyNDAzODU4OSwicGF0aCI6Ii8xMDM2NjE2NTgvMzU4OTY3NDczLTMyMmRiYzlhLWE0ODUtNDcxMy05YTk0LTI2NWJmNDM1OWEzOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOVQwMzM2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzljMDY2MTdiYmNkYjYzY2MyYzA1MDYyNjNmZTQxMzNlYmRhNDYzNDQ5NTM0NzIyNWFjMTUwZTczNjY1Nzc5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ly7YgwtnRDd5U6vfAlYZtlvQuxO_NwiXLgnk4blkdoQ)

这图的菜单结构完全不一样，可能 B 站又搞了新东西，我还没遇到

**Reply by WakelessSloth56** [reply]
新菜单里已经找不到 AI 屏蔽的设置了，甚至连等级都调不了？
修改 `window.webAbTest.danmuku_block_version` 失效之后，还能通过重载 `player.__core().blockStore.aiJudge()` 方法来直接干掉 AI 屏蔽，但现在 `player.__core()` 直接不返回这些 xxx`Store` 了。
我也在找新的途径来干掉沟槽的 AI 屏蔽

**Reply by WakelessSloth56** [reply]
倒是有个最终绝招，直接拦截播放器的脚本然后替换文本……确实能生效但是就emmmm还是找到别的方法更好……

> 更新：找到了一个方法，见下一条评论

```javascript
new MutationObserver(async (mutations, observer) => {
    const script = mutations
      .flatMap(e => [...e.addedNodes])
      .filter(e => e.tagName === 'SCRIPT')
      .find(e => e.src.match(/core\..{8}\.js/) !== null);

    if (script) {
        observer.disconnect()
        script.remove()

        let text = await fetch(script.src).then(e => e.text())

        // 直接干掉 AI 过滤
        text = text.replace(/aiJudge=.*?\}/, 'aiJudge=function(e,t){return false}')

        const newScript = document.createElement('script')
        newScript.textContent = text
        document.querySelector('head').appendChild(newScript)
    }
}).observe(document, {
    childList: true,
    subtree: true,
})
```

**Comment by WakelessSloth56** [comment]
现在的 `aiLevel` 已经和所谓的“智能云屏蔽”没什么关系了，它居然是根据弹幕区域和弹幕密度来算的，迷惑…………
我说为什么不管怎么设置它都是 `2`…………
```javascript
/*! @metadata nano: 4.8.47-rc.7958.0-0a64e57c (2024-09-13T10:25:34+08:00) */

Object.defineProperty(t.prototype, 'aiLevel', {
    get: function () {
        return this.state.dmarea < 100 && this.state.dmarea > 0
            ? 3
            : this.state.dmdensity === n.most || this.state.dmdensity === n.more
            ? 2
            : this.state.dmdensity === n.normal
            ? 3
            : void 0;
    },
    enumerable: !1,
    configurable: !0,
});
```

于是，取消“智能云屏蔽”的可行方法如下：

```javascript
// ==UserScript==
// @name         禁用弹幕智能云屏蔽
// @namespace    https://github.com/the1812/Bilibili-Evolved/discussions/4828
// @version      2024-09-14
// @description  禁用弹幕智能云屏蔽
// @author       WakelessSloth56
// @match        https://www.bilibili.com/video/*
// @grant        none
// @run-at       document-start
// ==/UserScript==
(function () {
    "use strict";
    const interval = setInterval(() => {
        if (window.player) {
            window.player.on(
                window.nano.EventType.Player_Play,
                () => {
                    // 每次播放时把弹幕密度设置为 0
                    const dmSetting = JSON.parse(localStorage.bpx_player_profile).dmSetting;
                    dmSetting.dmdensity = 0;
                    window.player.setDanmakuSetting(dmSetting);
                }
            );
            clearInterval(interval);
        }
    }, 100);
})();
```

**Reply by KuiPer-LTW** [reply]
这个和上面那个老哥的是二选一对吧，就看自己的弹幕选项是什么样式的

**Reply by WakelessSloth56** [reply]
对。要是仅仅是没有启用的复选框但是能调屏蔽等级，用上面那个；连等级都调不了的用这个。

**Reply by GalaxySnail** [reply]
我今天也被灰度到了，但是使用这个脚本仍然有一些弹幕无法显示

**Reply by KuiPer-LTW** [reply]
> 我今天也被灰度到了，但是使用这个脚本仍然有一些弹幕无法显示

我今天不知道为什么也不行了，经过测试发现把它俩一起打开又OK了，你试试
顺便也问问大佬我这种情况是个例还是怎么回事，如果可以复现那能不能把这俩脚本合并到一起？

**Reply by WakelessSloth56** [reply]
我新开了一个讨论 https://github.com/the1812/Bilibili-Evolved/discussions/4920 ，后续我会继续关注这个沟槽玩意╮(╯-╰)╭

**Reply by GalaxySnail** [reply]
> 我今天不知道为什么也不行了，经过测试发现把它俩一起打开又OK了，你试试

经我尝试两个脚本不论是否一起打开，总有一些弹幕无法显示，一起打开也不能解决，我也弄不明白原因（

**Reply by WakelessSloth56** [reply]
@KuiPer-LTW 把这个讨论关掉（outdated）吧，后续在新的讨论 #4920 继续。
